README.md
shadowsocks
===========

[![PyPI version]][PyPI]
[![Build Status]][Travis CI]

A fast tunnel proxy that helps you bypass firewalls.

Features:
- TCP & UDP support
- User management API
- TCP Fast Open
- Workers and graceful restart
- Destination IP blacklist

Server
----------

### Install

Debian / Ubuntu:

    apt-get install python-pip
    pip install git+https://github.com/shadowsocks/shadowsocks.git@master

CentOS:

    yum install python-setuptools && easy_install pip
    pip install git+https://github.com/shadowsocks/shadowsocks.git@master

Windows:

See [Install Shadowsocks Server on Windows](https://github.com/shadowsocks/shadowsocks/wiki/Install-Shadowsocks-Server-on-Windows).

### Usage

    ssserver -p 443 -k password -m aes-256-cfb

To run in the background:

    sudo ssserver -p 443 -k password -m aes-256-cfb --user nobody -d start

To stop:

    sudo ssserver -d stop

To check the log:

    sudo less /var/log/shadowsocks.log

Check all the options via `-h`. You can also use a [Configuration] file
instead.

### Usage with Config File

[Create configeration file and run](https://github.com/shadowsocks/shadowsocks/wiki/Configuration-via-Config-File)

To start:

    ssserver -c /etc/shadowsocks.json


Documentation
-------------

You can find all the documentation in the [Wiki](https://github.com/shadowsocks/shadowsocks/wiki).

License
-------

Apache License







[Build Status]:      https://img.shields.io/travis/shadowsocks/shadowsocks/master.svg?style=flat
[PyPI]:              https://pypi.python.org/pypi/shadowsocks
[PyPI version]:      https://img.shields.io/pypi/v/shadowsocks.svg?style=flat
[Travis CI]:         https://travis-ci.org/shadowsocks/shadowsocks



CHANGES
3.3.3 2017-06-03
- add DNS cache
- add tls1.2_ticket_fastauth
- fix bugs

3.3.2 2017-05-20
- revert http reply
- refine tls1.2_ticket_auth error detector

3.3.1 2017-05-18
- fix stop script
- Async DNS query under UDP
- fix old version of OpenSSL
- http reply

3.3.0 2017-05-11
- connect_log include local addr & port
- fix auth_chain_a UDP bug
- add "additional_ports_only"
- add interface legendsockssr
- run with newest python version
- parse comment in hosts
- update mujson_mgr
- add cymysql setup script
- new speed tester
- fix leaks
- bugs fixed

3.2.0 2017-04-27
- add auth_chain_a
- remove auth_aes128, auth_sha1, auth_sha1_v2, verify_simple, auth_simple, verify_sha1

3.1.2 2017-04-07
- display UID
- auto adjust TCP MSS

3.1.1 2017-03-25
- add "New session ticket"
- ignore bind 10.0.0.0/8 and 192.168.0.0/16 by default
- improve rand size under auth_aes128_*
- fix bugs

3.1.0 2017-03-16
- add "glzjinmod" interface
- rate limit
- add additional_ports in config

3.0.4 2017-01-08
- multi-user in single port

3.0.1 2017-01-03
- remove auth_aes128_*_compatible

3.0.0 2016-12-23
- http_simple fix bugs
- tls1.2_ticket_auth fix bug & defaule time diff set to 86400s

2.9.7 2016-11-22
- manage client with LRUCache
- catch bind error
- fix import error of resource on windows
- print RLIMIT_NOFILE
- always close cymysql objects
- add init script

2.9.6 2016-10-17
- tls1.2_ticket_auth random packet size

2.9.5.1 2016-10-16
- UDP bind address

2.9.5 2016-10-13
- add auth_aes128_md5 and auth_aes128_sha1

2.9.4 2016-10-11
- sync client version

2.6.13 2015-11-02
- add protocol setting

2.6.12 2015-10-27
- IPv6 first
- Fix mem leaks
- auth_simple plugin
- remove FORCE_NEW_PROTOCOL
- optimize code

2.6.11 2015-10-20
- Obfs plugin
- Obfs parameters
- UDP over TCP
- TCP over UDP (experimental)
- Fix socket leaks
- Catch abnormal UDP package

2.6.10 2015-06-08
- Optimize LRU cache
- Refine logging

2.6.9 2015-05-19
- Fix a stability issue on Windows

2.6.8 2015-02-10
- Support multiple server ip on client side
- Support --version
- Minor fixes

2.6.7 2015-02-02
- Support --user
- Support CIDR format in --forbidden-ip
- Minor fixes

2.6.6 2015-01-23
- Fix a crash in forbidden list

2.6.5 2015-01-18
- Try both 32 bit and 64 bit dll on Windows

2.6.4 2015-01-14
- Also search lib* when searching libraries

2.6.3 2015-01-12
- Support --forbidden-ip to ban some IP, i.e. localhost
- Search OpenSSL and libsodium harder
- Now works on OpenWRT

2.6.2 2015-01-03
- Log client IP

2.6.1 2014-12-26
- Fix a problem with TCP Fast Open on local side
- Fix sometimes daemon_start returns wrong exit status

2.6 2014-12-21
- Add daemon support

2.5 2014-12-11
- Add salsa20 and chacha20

2.4.3 2014-11-10
- Fix an issue on Python 3
- Fix an issue with IPv6

2.4.2 2014-11-06
- Fix command line arguments on Python 3
- Support table on Python 3
- Fix TCP Fast Open on Python 3

2.4.1 2014-11-01
- Fix setup.py for non-utf8 locales on Python 3

2.4 2014-11-01
- Python 3 support
- Performance improvement
- Fix LRU cache behavior

2.3.2 2014-10-11
- Fix OpenSSL on Windows

2.3.1 2014-10-09
- Does not require M2Crypto any more

2.3 2014-09-23
- Support CFB1, CFB8 and CTR mode of AES
- Do not require password config when using port_password
- Use SIGTERM instead of SIGQUIT on Windows

2.2.2 2014-09-14
- Fix when multiple DNS set, IPv6 only sites are broken

2.2.1 2014-09-10
- Support graceful shutdown
- Fix some bugs

2.2.0 2014-09-09
- Add RC4-MD5 encryption

2.1.0 2014-08-10
- Use only IPv4 DNS server
- Does not ship config.json
- Better error message

2.0.12 2014-07-26
- Support -q quiet mode
- Exit 0 when showing help with -h

2.0.11 2014-07-12
- Prefers IP addresses over hostnames, more friendly with socksify and openvpn

2.0.10 2014-07-11
- Fix UDP on local

2.0.9 2014-07-06
- Fix EWOULDBLOCK on Windows
- Fix Unicode config problem on some platforms

2.0.8 2014-06-23
- Use multiple DNS to query hostnames

2.0.7 2014-06-21
- Fix fastopen on local
- Fallback when fastopen is not available
- Add verbose logging mode -vv
- Verify if hostname is valid

2.0.6 2014-06-19
- Fix CPU 100% on POLL_HUP
- More friendly logging

2.0.5 2014-06-18
- Support a simple config format for multiple ports

2.0.4 2014-06-12
- Fix worker master

2.0.3 2014-06-11
- Fix table encryption with UDP

2.0.2 2014-06-11
- Add asynchronous DNS in TCP relay

2.0.1 2014-06-05
- Better logging
- Maybe fix bad file descriptor

2.0 2014-06-05
- Use a new event model
- Remove gevent
- Refuse to use default password
- Fix a problem when using multiple passwords with table encryption

1.4.5 2014-05-24
- Add timeout in TCP server
- Close sockets in master process

1.4.4 2014-05-17
- Support multiple workers

1.4.3 2014-05-13
- Fix Windows

1.4.2 2014-05-10
- Add salsa20-ctr cipher

1.4.1 2014-05-03
- Fix error log
- Fix EINPROGESS with some version of gevent

1.4.0 2014-05-02
- Adds UDP relay
- TCP fast open support on Linux 3.7+

1.3.7 2014-04-10
- Fix a typo in help

1.3.6 2014-04-10
- Fix a typo in help

1.3.5 2014-04-07
- Add help
- Change default local binding address into 127.0.0.1

1.3.4 2014-02-17
- Fix a bug when no config file exists
- Client now support multiple server ports and multiple server/port pairs
- Better error message with bad config.json format and wrong password

1.3.3 2013-07-09
- Fix default key length of rc2

1.3.2 2013-07-04
- Server will listen at server IP specified in config
- Check config file and show some warning messages

1.3.1 2013-06-29
- Fix -c arg

1.3.0 2013-06-22
- Move to pypi

1.2.3 2013-06-14
- add bind address

1.2.2 2013-05-31
- local can listen at ::0 with -6 arg; bump 1.2.2

1.2.1 2013-05-23
- Fix an OpenSSL crash

1.2 2013-05-22
- Use random iv, we finally have strong encryption

1.1.1 2013-05-21
- Add encryption, AES, blowfish, etc.

1.1 2013-05-16
- Support IPv6 addresses (type 4)
- Drop Python 2.5 support

1.0 2013-04-03
- Fix -6 IPv6

0.9.4 2013-03-04
- Support Python 2.5

0.9.3 2013-01-14
- Fix conn termination null data


encrypt.py
#!/usr/bin/env python

# Copyright (c) 2012 clowwindy
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import sys
import hashlib
import string
import struct
import logging


def random_string(length):
    import M2Crypto.Rand
    return M2Crypto.Rand.rand_bytes(length)


def get_table(key):
    m = hashlib.md5()
    m.update(key)
    s = m.digest()
    (a, b) = struct.unpack('<QQ', s)
    table = [c for c in string.maketrans('', '')]
    for i in xrange(1, 1024):
        table.sort(lambda x, y: int(a % (ord(x) + i) - a % (ord(y) + i)))
    return table

encrypt_table = None
decrypt_table = None


def init_table(key, method=None):
    if method == 'table':
        method = None
    if method:
        try:
            __import__('M2Crypto')
        except ImportError:
            logging.error('M2Crypto is required to use encryption other than default method')
            sys.exit(1)
    if not method:
        global encrypt_table, decrypt_table
        encrypt_table = ''.join(get_table(key))
        decrypt_table = string.maketrans(encrypt_table, string.maketrans('', ''))
    else:
        try:
            Encryptor(key, method)  # make an Encryptor to test if the settings if OK
        except Exception as e:
            logging.error(e)
            sys.exit(1)


def EVP_BytesToKey(password, key_len, iv_len):
    # equivalent to OpenSSL's EVP_BytesToKey() with count 1
    # so that we make the same key and iv as nodejs version
    # TODO: cache the results
    m = []
    i = 0
    while len(''.join(m)) < (key_len + iv_len):
        md5 = hashlib.md5()
        data = password
        if i > 0:
            data = m[i - 1] + password
        md5.update(data)
        m.append(md5.digest())
        i += 1
    ms = ''.join(m)
    key = ms[:key_len]
    iv = ms[key_len:key_len + iv_len]
    return (key, iv)


method_supported = {
    'aes-128-cfb': (16, 16),
    'aes-192-cfb': (24, 16),
    'aes-256-cfb': (32, 16),
    'bf-cfb': (16, 8),
    'camellia-128-cfb': (16, 16),
    'camellia-192-cfb': (24, 16),
    'camellia-256-cfb': (32, 16),
    'cast5-cfb': (16, 8),
    'des-cfb': (8, 8),
    'idea-cfb': (16, 8),
    'rc2-cfb': (8, 8),
    'rc4': (16, 0),
    'seed-cfb': (16, 16),
}


class Encryptor(object):
    def __init__(self, key, method=None):
        if method == 'table':
            method = None
        self.key = key
        self.method = method
        self.iv = None
        self.iv_sent = False
        self.cipher_iv = ''
        self.decipher = None
        if method is not None:
            self.cipher = self.get_cipher(key, method, 1, iv=random_string(32))
        else:
            self.cipher = None

    def get_cipher_len(self, method):
        method = method.lower()
        m = method_supported.get(method, None)
        return m

    def iv_len(self):
        return len(self.cipher_iv)

    def get_cipher(self, password, method, op, iv=None):
        import M2Crypto.EVP
        password = password.encode('utf-8')
        method = method.lower()
        m = self.get_cipher_len(method)
        if m:
            key, iv_ = EVP_BytesToKey(password, m[0], m[1])
            if iv is None:
                iv = iv_[:m[1]]
            if op == 1:
                self.cipher_iv = iv[:m[1]]  # this iv is for cipher, not decipher
            return M2Crypto.EVP.Cipher(method.replace('-', '_'), key, iv, op, key_as_bytes=0, d='md5', salt=None, i=1, padding=1)

        logging.error('method %s not supported' % method)
        sys.exit(1)

    def encrypt(self, buf):
        if len(buf) == 0:
            return buf
        if self.method is None:
            return string.translate(buf, encrypt_table)
        else:
            if self.iv_sent:
                return self.cipher.update(buf)
            else:
                self.iv_sent = True
                return self.cipher_iv + self.cipher.update(buf)

    def decrypt(self, buf):
        if len(buf) == 0:
            return buf
        if self.method is None:
            return string.translate(buf, decrypt_table)
        else:
            if self.decipher is None:
                decipher_iv_len = self.get_cipher_len(self.method)[1]
                decipher_iv = buf[:decipher_iv_len]
                self.decipher = self.get_cipher(self.key, self.method, 0, iv=decipher_iv)
                buf = buf[decipher_iv_len:]
                if len(buf) == 0:
                    return buf
            return self.decipher.update(buf)
